### PR TITLE
[WIP] win10-1803: add version and syscall numbers

### LINKED ISSUE #3090: Windows 10 1803 support?
Would work on Windows 1803 support be welcome? I'm not a Windows kernel person, but I should be able to mostly follow the work in #2698 and the updated syscall table here: https://github.com/tinysec/windows-syscall-table/blob/master/10.0.17134-sp0-windows-10-rs4-1803/ssdt.md.

I'll probably need some help with `NtAllocateVirtualMemoryEx`, since that probably needs to be intercepted like `NtAllocateVirtualMemory` currently is.

### PR DESCRIPTION
**Pull request by woodruffw**
I still need to add the code that installs the pre- and post-hooks for `NtAllocateVirtualMemoryEx` (and might need some help with that), but this contains most of the changes.

Closes #3090.

### CODE REVIEW

File: core/win32/syscallx.h [hunk 1]
```diff
@@ -98,59 +98,60 @@
...
+SYSCALL(QueryVirtualMemory,           true,     6, 0x18, 0x81, 0x81, 0x81, 0x9c, 0x0b2,    0, 0x020, 0x0ba, 0x0fd, 0x020, 0x0fd, 0x020, 0x10b,0x020, 0x08f,0x00021,0x021, 0x092,0x00022,0x022, 0x094,0x00023,0x023, 0x094,0x00023,0x023, 0x095,0x00023,0x023, 0x096,0x00023,0x023, 0x097,0x00023,0x023, 0x097,0x00023,0x023)
+SYSCALL(WriteVirtualMemory,           true,     5, 0x14, 0xc9, 0xcb, 0xcb, 0xf0, 0x115,    0, 0x037, 0x11f, 0x16a, 0x037, 0x166, 0x037, 0x18f,0x037, 0x002,0x00038,0x038, 0x003,0x00039,0x039, 0x004,0x0003a,0x03a, 0x004,0x0003a,0x03a, 0x004,0x0003a,0x03a, 0x004,0x0003a,0x03a, 0x004,0x0003a,0x03a, 0x004,0x0003a,0x03a)
+SYSCALL(MapViewOfSection,             true,    10, 0x28, 0x49, 0x49, 0x49, 0x5d, 0x06c,    0, 0x025, 0x071, 0x0b1, 0x025, 0x0b1, 0x025, 0x0a8,0x025, 0x0f3,0x00026,0x026, 0x0f6,0x00027,0x027, 0x0fa,0x00028,0x028, 0x0fa,0x00028,0x028, 0x0fc,0x00028,0x028, 0x0fe,0x00028,0x028, 0x101,0x00028,0x028, 0x101,0x00028,0x028)
```
Comment (derekbruening) [inline core/win32/syscallx.h]: NtMapViewOfSectionEx was also added and we'll need to handle it as well.

File: core/lib/instrument.c [hunk 2]
```diff
@@ -2712,6 +2712,7 @@ dr_get_os_version(dr_os_version_info_t *info)
     get_os_version_ex(&ver, &sp_major, &sp_minor);
```
Comment (derekbruening) [inline core/lib/instrument.c]: The api/docs/release.dox changelog message is missing: maybe you planned to add it once NtAllocateVirtualMemoryEx and NtMapViewOfSectionEx support is in?  I would say, add it here in the same diff that raises max_supported_os_version.

**Comment by woodruffw** [comment]
> If we consider NtAllocateVirtualMemoryEx and NtMapViewOfSectionEx part of #3090, the commit message should just say "Issue: #3090" rather than that it fixes #3090 (also note that we use Fixes rather than Closes).

Got it, I'll make the changes above and amend the message.

File: core/win32/syscall.c [hunk 3]
```diff
@@ -113,12 +113,42 @@ int windows_unknown_syscalls[TRAMPOLINE_MAX];
  * planned Dr. Memory route (DrMi#1848) and store these numbers in a separate file
  * that is updated via a separate standalone utility run once by the user.
  */
+SYS_CONST int windows_10_1803_x64_syscalls[TRAMPOLINE_MAX] = {
+#define SYSCALL(name, act, nargs, arg32, ntsp0, ntsp3, ntsp4, w2k, xp, wow64, xp64,\
+                w2k3, vista0, vista0_x64, vista1, vista1_x64, w7x86, w7x64,        \
+                w8x86, w8w64, w8x64, w81x86, w81w64, w81x64, w10x86, w10w64, w10x64,\
+                w11x86, w11w64, w11x64, w12x86, w12w64, w12x64, w13x86, w13w64, w13x64, \
+                w14x86, w14w64, w14x64, w15x86, w15w64, w15x64) \
+    w14x64,
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply core/win32/syscall.c]: I'm going to run a build on 1803 later this afternoon. I hadn't done so yet, which is why I hadn't caught this -- it does cause a debug message as-is, though.

**Comment by woodruffw** [comment]
> FYI, we may put in place a big shift to use clang-format everywhere (including reformatting existing files) this weekend

Thanks for the heads-up!

**Comment by woodruffw** [comment]
I can't find any authoritative information on the parameters taken by `NtAllocateVirtualMemoryEx` and `NtMapViewOfSectionEx`, but [this repo]() contains some signatures:

```c
NTSTATUS
NTAPI
NtAllocateVirtualMemoryEx(
    IN HANDLE ProcessHandle,
    IN OUT PPVOID lpAddress,
    IN ULONG_PTR ZeroBits,
    IN OUT PSIZE_T pSize,
    IN ULONG flAllocationType,
    IN OUT PVOID DataBuffer OPTIONAL,
    IN ULONG DataCount
    );
```

```c
NTSTATUS
NTAPI
NtMapViewOfSectionEx(
    IN HANDLE SectionHandle,
    IN HANDLE ProcessHandle,
    IN OUT PLARGE_INTEGER SectionOffset,
    IN OUT PPVOID BaseAddress,
    IN OUT PSIZE_T ViewSize,
    IN ULONG AllocationType,
    IN ULONG Protect,
    IN OUT PVOID DataBuffer OPTIONAL,
    IN ULONG DataCount
    );
```

In particular, it looks like the protection flag has disappeared versus `NtAllocateVirtualMemory`, and that the `InheritDisposition` parameter has disappeared versus `NtMapViewOfSection`. Both add two new paramers: `DataBuffer` and `DataCount`, and I'm not sure what they do (yet).

Any thoughts on how to approach this? Information on these two is pretty scarce.

**Comment by derekbruening** [comment]
I would say, separate the details out: for now, just put in an empty handler with a visible message (debug-only seems fine, so ASSERT_CURIOSITY or the lighter SYSLOG_INTERNAL_WARNING or sthg) so we know if we hit these, and get the rest of the patch in.  Add the full handling of these 2 in a separate patch.

**Comment by derekbruening** [comment]
Put a `FIXME i#3090: ....` comment in the handlers.

**Comment by derekbruening** [comment]
As for figuring out what those syscalls do, if you can find some app that actually invokes them it is much easier of course -- which the debug message will help with.

**Comment by derekbruening** [comment]
Search the higher-level libraries (win32 API) for invocations.

**Comment by derekbruening** [comment]
Please don't force-push: now the reviewer can't see just the new changes.  Use separate commits: don't change the history on a shared branch.  The final squash-and-merge into master will clean it up.  This should be spelled out on the wiki pages.

**Comment by woodruffw** [comment]
Oh, just saw that. Sorry about the pushes, I'll keep the history from here on.

File: core/win32/syscall.c [hunk 4]
```diff
@@ -4288,6 +4367,9 @@ void post_system_call(dcontext_t *dcontext)
         postsys_MapViewOfSection(dcontext, param_base, success);
         KSTOP(post_syscall_map);
```
Comment (derekbruening) [inline core/win32/syscall.c]: We probably want the same KSTART..KSTOP for Ex but we could move it to some shared handler once we have Ex handling: they should share a lot of code.

**Comment by derekbruening** [comment]
run arm tests

File: core/win32/syscall.c (hunk 4 above)
Comment (woodruffw) [reply core/win32/syscall.c]: Yeah, I was wondering about those. I'll add them in the subsequent PR if the handlers are too different (but they shouldn't be, as you said).

**Comment by derekbruening** [comment]
Merged.  Thank you for the patch.

**Pull request merged**